langmeta/shared/src/main/scala/org/langmeta/semanticdb/internal/package.scala

```diff
@@ -25,94 +30,128 @@ package object semanticdb {
-                    .find(RelativePath(sfilename))
-                    .getOrElse(sys.error(s"can't find $sfilename in $sourcepath"))
-              dInput.File(AbsolutePath(uri.getPath))
-            } else {
-              dInput.VirtualFile(sfilename.toString, scontents)
-            }
-          }
-          object sPosition {
-            def unapply(spos: s.Position): Option[dPosition] = {
-              Some(dPosition.Range(dinput, spos.start, spos.end))
-            }
-          }
-          object sSeverity {
-            def unapply(sseverity: s.Message.Severity): Option[d.Severity] = {
-              sseverity match {
-                case s.Message.Severity.INFO => Some(d.Severity.Info)
-                case s.Message.Severity.WARNING => Some(d.Severity.Warning)
-                case s.Message.Severity.ERROR => Some(d.Severity.Error)
-                case _ => None
-              }
-            }
-          }
-          object sResolvedSymbol {
-            def unapply(sresolvedsymbol: s.ResolvedSymbol): Option[d.ResolvedSymbol] = sresolvedsymbol match {
-              case s.ResolvedSymbol(d.Symbol(dsym), Some(s.Denotation(dflags, dname: String, dsignature: String, snames, smembers))) =>
-                val ddefninput = dInput.Denotation(dsignature, dsym)
-                val dnames = snames.toIterator.map {
-                  case s.ResolvedName(Some(s.Position(sstart, send)), d.Symbol(dsym), disDefinition) =>
-                    val ddefnpos = dPosition.Range(ddefninput, sstart, send)
-                    d.ResolvedName(ddefnpos, dsym, disDefinition)
-                  case other =>
-                    sys.error(s"bad protobuf: unsupported name $other")
-                }.toList
-                val dmembers: Option[List[d.Signature]] =
-                  if (smembers.isEmpty) None
-                  else Some(
-                    smembers.toIterator.map { smember =>
-                      if (smember.endsWith("#")) d.Signature.Type(smember.stripSuffix("#"))
-                      else if (smember.endsWith(".")) d.Signature.Term(smember.stripSuffix("."))
-                      else sys.error(s"Unexpected signature $smember")
-                    }.toList
-                  )
-                val ddefn = d.Denotation(dflags, dname, dsignature, dnames, dmembers)
-                Some(d.ResolvedSymbol(dsym, ddefn))
-              case other => sys.error(s"bad protobuf: unsupported denotation $other")
-            }
+    private def sdocumentsMerged: Seq[s.Document] = {
+      def isOnlyMessages(sdocument: s.Document) =
+        sdocument.contents.isEmpty &&
+          sdocument.messages.nonEmpty &&
+          sdocument.names.isEmpty &&
+          sdocument.synthetics.isEmpty &&
+          sdocument.symbols.isEmpty
+      if (sdatabase.documents.length <= 1) {
+        // NOTE(olafur) the most common case is that there is only a single database
+        // per document so we short-circuit here if that's the case.
```

Comment:
How about merging the databases if the input is the same? That'll be more general than hardcoding this very specific condition.

Reply:
This method is a relic of a failed attempt, I've removed it.

> if the input is the same

Message-only documents have no .contents, and s.Database has no input.

Reply:
But it does have an input, doesn't it? `input = m.Input.File(unit.source.file.file)`. Can we use this to correlate documents?

Reply:
If we don't have time to do this right now, let's merge but at least open an issue to make sure we follow through next time. This is some technical debt that we're taking on.

Reply:
We don't have a source root so `m.Input.File()` won't work unless `sys.props("user.dir")` coincidentally matches --sourceroot. I think a step in the right direction to make this less hacky is https://github.com/scalameta/scalameta/issues/1151 we can persist the flavor of the persisted message in `s.Document`, one field could be to indicate that this is a message-only Document. Message-only are relevant outside of this use-case, I would use this also to implement something like https://github.com/olafurpg/sbt-messagehost